fix(a11y): type error on focus initialization

znpopov opened the pull request:
The key-focus component does not handle properly the focus initialization, when some of the focusable elements are being removed.
In case the focus was on an element, whose index is no longer available, the component throws a type error.
This change handles the case by initializing the focus on the last focusable element.

Signed-off-by: Zdravko Popov <zpopov@vmware.com>

## PR Checklist

Please check if your PR fulfills the following requirements:

* [x] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

* [x] Bugfix
* [ ] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] clarity.design website / infrastructure changes
* [ ] Other... Please describe:

## What is the current behavior?
https://stackblitz.com/edit/clarity-dark-theme-v2-zushxr?file=src%2Fapp%2Fapp.component.ts

![TypeError](https://user-images.githubusercontent.com/26929093/74756774-c9e77400-527d-11ea-88b1-3807435915c4.png)

Issue Number: N/A

## What is the new behavior?
The focus is initialized on the last focusable element.

## Does this PR introduce a breaking change?

* [ ] Yes
* [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
This bug was introduced with the addition of the key-focus component.
For this reason, the fix needs to be back-ported to v2 and v1 branches.

gnomeontherun commented on src/clr-angular/utils/focus/key-focus/key-focus.ts:
```diff
@@ -108,7 +108,7 @@ export class ClrKeyFocus {
   private get currentItem() {
     if (this._current >= this.focusableItems.length) {
-      return null;
+      this._current = 0;
```
Why should the focus move to the first item? This might make sense if you are on the first item and it is removed, but if you are on the last item and it is removed I would think focus should move to the new last focusable item? 

Your demo shows the latter behavior, so I would expect focus to move to the `this._current = this.focusableItems.length - 1`. 

If any item in the list is removed except the last item, the current index remains the same and items shift over so that shouldn't be an issue.

znpopov replied:
Thanks for the review.
I have updated the request with the corresponding changes.

znpopov commented:
It turns out that the key-focus component can be used by two different ways:
1. Providing a list of focusable elements.
2. If the list is not provided, the component  auto queries for clrKeyFocusItem.

I have addressed the requested changes for both cases and added corresponding tests.

gnomeontherun commented on src/clr-angular/utils/focus/key-focus/key-focus.spec.ts:
```diff
@@ -280,7 +292,19 @@ describe('KeyFocus directive', () => {
-      expect(clarityDirective.focusableItems.length).toBe(2);
+      expect(clarityDirective.focusableItems.length).toBe(3);
+    });
+    it('current value updates, when elements are being removed', () => {
```
This is just a duplicate test of the above using the template instead of the API. It can be removed.

znpopov replied:
It is not an exact duplicate. 
In this scenario, the changes are handled by the `focusableItems` setter.
In the case of the API test, the changes are handled by the `listenForItemUpdates()` method.
In my opinion, both tests are needed.

gnomeontherun replied:
What is triggering both tests to actually change is simply the toggling of `showLast`, which triggers the `listenForItemUpdates` to fire based on the QueryList change. These tests are just setting up the menu in two different ways ,but both asserting the result the same way (hence the duplication). 

Thinking about it more, this test should not be asserting that `current` is set but that the `document.activeElement` equals the expected button. If you change the test to assert that, I would be satisfied that its testing two different ways.

znpopov replied:
Thanks. The test has been updated with the suggested assertions.

znpopov commented on src/clr-angular/utils/focus/key-focus/key-focus.ts:
```diff
@@ -108,7 +108,7 @@ export class ClrKeyFocus {
   private get currentItem() {
     if (this._current >= this.focusableItems.length) {
-      return null;
+      this._current = this.focusableItems.length - 1;
```
Thanks for this.

I moved the logic into the `initializeFocus()` method, since changes can be handled from the `listenForItemUpdates()` method or from the `focusableItems` setter.

With this logic, if all items are removed temporarily, the getter will never be called.

znpopov commented:
Hi @gnomeontherun,
Do you need more input from my side, in order the PR to be merged?
Thanks!
